qutip/cy/td_qobj_cy.pyx

```diff
@@ -0,0 +1,583 @@
+
+    def call(self, double t, int data=0):
+        cdef CSR_Matrix out
+        out.is_set = 0
+        copy_CSR(&out, &self.cte)
+        scipy_obj = CSR_to_scipy(&out)
+        # free_CSR(&out)? data is own by the scipy_obj?
+        if data:
+            return scipy_obj
+        else:
+            return Qobj(scipy_obj,dims = self.dims)
+
+    def call_with_coeff(self, double t, complex[::1] coeff, int data=0):
+        cdef CSR_Matrix out
+        out.is_set = 0
+        copy_CSR(&out, &self.cte)
+        scipy_obj = CSR_to_scipy(&out)
+        # free_CSR(&out)? data is own by the scipy_obj?
+        if data:
+            return scipy_obj
+        else:
+            return Qobj(scipy_obj)
+
+    def get_ptr(self):
+        return PyLong_FromVoidPtr(<void*> self)
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.cdivision(True)
+    cdef void _rhs_mat(self, double t, complex* vec, complex* out):
+        spmvpy(self.cte.data, self.cte.indices, self.cte.indptr, vec, 1.,
+               out, self.shape0)
+
+    def rhs(self, double t, complex[::1] vec):
+        cdef np.ndarray[complex, ndim=1] out = np.zeros(self.shape0,
+                                                        dtype=complex)
+        self._rhs_mat(t, &vec[0], &out[0])
+        return out
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.cdivision(True)
+    cdef complex _expect_mat(self, double t, complex* vec, int isherm):
+        cdef complex[::1] y = np.zeros(self.shape0, dtype=complex)
+        spmvpy(self.cte.data, self.cte.indices, self.cte.indptr, vec, 1.,
+               &y[0], self.shape0)
+        cdef int row
+        cdef complex dot = 0
+        for row from 0 <= row < self.shape0:
+            dot += conj(vec[row])*y[row]
+        if isherm:
+            return real(dot)
+        else:
+            return dot
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.cdivision(True)
+    cdef complex _expect_mat_super(self, double t, complex* vec, int isherm):
+        cdef int row
+        cdef int jj, row_start, row_end
+        cdef int num_rows = self.shape0
+        cdef int n = <int>libc.math.sqrt(num_rows)
+        cdef complex dot = 0.0
+
+        for row from 0 <= row < num_rows by n+1:
+            row_start = self.cte.indptr[row]
+            row_end = self.cte.indptr[row+1]
+            for jj from row_start <= jj < row_end:
+                dot += self.cte.data[jj]*vec[self.cte.indices[jj]]
+
+        if isherm:
+            return real(dot)
+        else:
+            return dot
+
+    def expect(self, double t, complex[::1] vec, int isherm):
+        if self.super:
+            return self._expect_mat_super(t, &vec[0], isherm)
+        else:
+            return self._expect_mat(t, &vec[0], isherm)
+
+
+cdef class cy_td_qobj(cy_qobj):
+    def __init__(self):
+        self.N_ops = 0
+        self.ops = <CSR_Matrix**> malloc(0 * sizeof(CSR_Matrix*))
```

Comment:
Calls to `malloc` should be replaced with `PyDataMem_NEW(size_t size)` from `numpy/arrayobject.h`. This is because the `malloc` called by the complied Cython code may not be the same as the `malloc` used when building Python.  For example, using GCC to compile QuTiP on Win.  Freeing data should also be done with ` PyDataMem_FREE(void * ptr)`.  Learned this the hard way.

Reply:
I will make the replacement this week.